bleak/_api.py

```diff
@@ -0,0 +1,117 @@
+        BaseBleakClient as _BleakClientImplementation,
+    )  # noqa: F401
+elif os.environ.get("P4A_BOOTSTRAP") is not None:
+    from bleak.backends.p4android.scanner import (
+        BleakScannerP4Android as _BleakScannerImplementation,
+    )  # noqa: F401
+    from bleak.backends.p4android.client import (
+        BleakClientP4Android as _BleakClientImplementation,
+    )  # noqa: F401
+elif platform.system() == "Linux":
+    if not _on_ci and not check_bluez_version(5, 43):
+        raise BleakError("Bleak requires BlueZ >= 5.43.")
+
+    from bleak.backends.bluezdbus.scanner import (
+        BleakScannerBlueZDBus as _BleakScannerImplementation,
+    )  # noqa: F401
+    from bleak.backends.bluezdbus.client import (
+        BleakClientBlueZDBus as _BleakClientImplementation,
+    )  # noqa: F401
+elif platform.system() == "Darwin":
+    try:
+        from CoreBluetooth import CBPeripheral  # noqa: F401
+    except Exception as ex:
+        raise BleakError("Bleak requires the CoreBluetooth Framework") from ex
+
+    from bleak.backends.corebluetooth.scanner import (
+        BleakScannerCoreBluetooth as _BleakScannerImplementation,
+    )  # noqa: F401
+    from bleak.backends.corebluetooth.client import (
+        BleakClientCoreBluetooth as _BleakClientImplementation,
+    )  # noqa: F401
+
+elif platform.system() == "Windows":
+    # Requires Windows 10 Creators update at least, i.e. Window 10.0.16299
+    _vtup = platform.win32_ver()[1].split(".")
+    if int(_vtup[0]) != 10:
+        raise BleakError(
+            "Only Windows 10 is supported. Detected was {0}".format(
+                platform.win32_ver()
+            )
+        )
+
+    if (int(_vtup[1]) == 0) and (int(_vtup[2]) < 16299):
+        raise BleakError(
+            "Requires at least Windows 10 version 0.16299 (Fall Creators Update)."
+        )
+
+    from bleak.backends.winrt.scanner import (
+        BleakScannerWinRT as _BleakScannerImplementation,
+    )  # noqa: F401
+    from bleak.backends.winrt.client import (
+        BleakClientWinRT as _BleakClientImplementation,
+    )  # noqa: F401
+
+else:
+    raise BleakError(f"Unsupported platform: {platform.system()}")
+
+# Now let's tie together the abstract class and the backend implementation
+class BleakScanner(_BleakScannerImplementation):
+    """
+    Interface for Bleak Bluetooth LE Scanners.
+
+    The actual implementation is dependent on the backend used, and some methods (notably the
+    constructor) may have additional optional arguments.
+
+
+    Args:
+        detection_callback:
+            Optional function that will be called each time a device is
+            discovered or advertising data has changed.
+        service_uuids:
+            Optional list of service UUIDs to filter on. Only advertisements
+            containing this advertising data will be received.
+        scanning_mode:
+            Set to "passive" to avoid the "active" scanning mode.
+    """
+
+    pass
```

Comment:
I was hoping that we could put all of the methods/properties here so all of the documentation would be in once place and not have to be repeated in the backends. And also so that when you jump to the definition of a method/property using an IDE you end up here.

Reply:
I haven't moved the documentation here, because frankly I'm still a bit out of my water with the whole ABC and typing stuff, and how I could document a method in the toplevel class (in an abstract method?) and then override the implementation from the subclass.

Reply:
I think this is resolved with the new `abstract_api.py`.